### PR TITLE
Issue #164 - refactor string unescaping

### PR DESCRIPTION
**Pull request by mkaminski1988**
This PR removes the string duplication (and string allocation) that occurs upon invoking `jsonpath()`, which was previously used to solve the [escaped string edge case](https://github.com/supermetrics-public/pecl-jsonpath/blob/main/tests/comparison_bracket_notation/025.phpt).

**Before**
`jsonpath()` duplicates the input string and passes it to the lexer. The lexer removes slashes from the string.

**After**
The lexer assigns tokens containing escaped characters the `LEX_LITERAL_UNESCAPED` type. When the parser encounters `LEX_LITERAL_UNESCAPED`, it runs the built-in `php_stripcslashes()` function on the allocated string.

I've also added a test that ensures string literal unescaping works. The existing tests in `tests/comparison_bracket_notation` only ensure that unescaping works on selectors.

### CODE REVIEW

File: src/jsonpath/lexer.c [hunk 1]
```diff
@@ -207,7 +206,6 @@ bool scan(char** p, struct jpath_token* token, char* json_path) {
         if (!extract_quoted_literal(p, json_path, token)) {
           return false;
         }
-        token->type = LEX_LITERAL;
```
Comment (mkaminski1988) [inline src/jsonpath/lexer.c:210]: This was redundant, already done in `extract_quoted_literal`

**Review by crocodele — APPROVED** [review]
:clap:
